src/mgmt/srv_drpc.c

```diff
@@ -1734,22 +1734,22 @@ pool_query_free_tier_stats(Mgmt__PoolQueryResp *resp)
 void
 ds_mgmt_drpc_pool_query(Drpc__Call *drpc_req, Drpc__Response *drpc_resp)
 {
-	struct drpc_alloc	alloc = PROTO_ALLOCATOR_INIT(alloc);
-	int			rc = 0;
-	Mgmt__PoolQueryReq	*req;
-	Mgmt__PoolQueryResp	resp = MGMT__POOL_QUERY_RESP__INIT;
-	Mgmt__StorageUsageStats	scm = MGMT__STORAGE_USAGE_STATS__INIT;
-	Mgmt__StorageUsageStats	nvme = MGMT__STORAGE_USAGE_STATS__INIT;
-	Mgmt__PoolRebuildStatus	rebuild = MGMT__POOL_REBUILD_STATUS__INIT;
-	uuid_t			uuid;
-	daos_pool_info_t	pool_info = {0};
-	d_rank_list_t		*svc_ranks;
-	d_rank_list_t		*ranks;
-	d_rank_range_list_t	*range_list;
-	char			*range_list_str = NULL;
-	bool			truncated;
-	size_t			len;
-	uint8_t			*body;
+	struct drpc_alloc       alloc = PROTO_ALLOCATOR_INIT(alloc);
+	int                     rc    = 0;
+	Mgmt__PoolQueryReq     *req;
+	Mgmt__PoolQueryResp     resp    = MGMT__POOL_QUERY_RESP__INIT;
+	Mgmt__StorageUsageStats scm     = MGMT__STORAGE_USAGE_STATS__INIT;
+	Mgmt__StorageUsageStats nvme    = MGMT__STORAGE_USAGE_STATS__INIT;
+	Mgmt__PoolRebuildStatus rebuild = MGMT__POOL_REBUILD_STATUS__INIT;
+	uuid_t                  uuid;
+	daos_pool_info_t        pool_info          = {0};
+	d_rank_list_t          *svc_ranks          = NULL;
+	d_rank_list_t          *enabled_ranks      = NULL;
+	d_rank_list_t          *disabled_ranks     = NULL;
+	char                   *enabled_ranks_str  = NULL;
+	char                   *disabled_ranks_str = NULL;
```

Comment:
I don't recall how many ranks can be in a pool. If it's a lot, the string may actually be the better way to transmit that value...

Reply:
Well, in the case of Aurora, a pool could in theory span 2048 ranks. As far as I'm aware there's no hard-coded limit on the number of ranks in a pool, so it could go higher for larger systems. 

You make a good point about the ranklist string being a more compact representation. As I said, I don't feel very strongly about this, I guess I was just reacting a little to what seemed like presentation-layer logic living in the drpc handler.

Reply:
From my side, I have to admit that I am also not found of using a string, and I will indeed prefer to use a list of intervals.
However, I will prefer to do that in a follow up PR with a dedicated ticket.
Does it makes sense.

Reply:
Lists of intervals might be a nice compromise here. If I'm understanding correctly, you're proposing something like:
  enabled_ranks: [0,7,9,15] (equivalent to "[0-7,9-15]")

Or, best-case:
  enabled_ranks: [0,15] (equivalent to "[0-15]")

Right? This would still be reasonably compact, and would still require some interpretation, but would at least eliminate the string manipulation steps for encode/decode.

Reply:
I do think we need to be careful about the interval list representation, if/when we attempt to implement it. How would we know the difference between intending ranks 0 and 15 vs 0-15?

I'm not crazy about using strings at this level myself. If we can come up with a better compact representation, I'd be happy to hear about it. :)

Reply:
> enabled_ranks: [0,15] (equivalent to "[0-15]")

I would find that confusing, and I imagine many admins would too.
Common nodeset / clush / pdsh syntax is
```
node[0,2] -> node0, node2
node[0-2] -> node0, node1, node2
```
And I wonder if GO even has a library for working with those?

Reply:
> > enabled_ranks: [0,15] (equivalent to "[0-15]")
> 
> I would find that confusing, and I imagine many admins would too. Common nodeset / clush / pdsh syntax is
> 
Well, to be clear, I'm not suggesting that these intervals would be exposed to admins. I think we're just discussing options for optimizing the protobuf messages sent between components. We would still want to pretty-print them as pdsh-style ranges for human consumption.

The current JSON output for these fields is an exploded list, which I think is correct. The JSON consumer shouldn't need to deal with range parsing or interval interpretation.

Reply:
> just discussing options for optimizing the protobuf messages

Gotcha. Sorry for misunderstanding :)

Reply:
> I do think we need to be careful about the interval list representation, if/when we attempt to implement it. How would we know the difference between intending ranks 0 and 15 vs 0-15?
> 
Yeah... I was still on my first cup of coffee when I proposed that. Should have known it couldn't be that simple. :)

Could probably do something like this to be less clever and more explicit:
```
+message RankInterval {
+  uint32 begin = 1;
+  uint32 end = 2;
+}
+
 // PoolQueryResp represents a pool query response.
 message PoolQueryResp {
        reserved 9;
@@ -229,8 +234,8 @@ message PoolQueryResp {
        repeated StorageUsageStats tier_stats = 8; // storage tiers usage stats
        uint32 version = 10; // latest pool map version
        uint32 leader = 11; // current raft leader (2.4)
-       string enabled_ranks = 12; // optional set of ranks enabled
-       string disabled_ranks = 13; // optional set of ranks disabled
+       repeated RankInterval enabled_ranks = 12; // optional set of ranks enabled
+       repeated RankInterval disabled_ranks = 13; // optional set of ranks disabled
```

That would map pretty easily into what we're doing in the [hostlist](https://github.com/daos-stack/daos/tree/master/src/control/lib/hostlist) package, and should be easier to work with on the C side than strings.

Reply:
@mjmac I like that idea.

Reply:
> > I do think we need to be careful about the interval list representation, if/when we attempt to implement it. How would we know the difference between intending ranks 0 and 15 vs 0-15?
> 
> Yeah... I was still on my first cup of coffee when I proposed that. Should have known it couldn't be that simple. :)
> 
> Could probably do something like this to be less clever and more explicit:
> 
> ```
> +message RankInterval {
> +  uint32 begin = 1;
> +  uint32 end = 2;
> +}
> +
>  // PoolQueryResp represents a pool query response.
>  message PoolQueryResp {
>         reserved 9;
> @@ -229,8 +234,8 @@ message PoolQueryResp {
>         repeated StorageUsageStats tier_stats = 8; // storage tiers usage stats
>         uint32 version = 10; // latest pool map version
>         uint32 leader = 11; // current raft leader (2.4)
> -       string enabled_ranks = 12; // optional set of ranks enabled
> -       string disabled_ranks = 13; // optional set of ranks disabled
> +       repeated RankInterval enabled_ranks = 12; // optional set of ranks enabled
> +       repeated RankInterval disabled_ranks = 13; // optional set of ranks disabled
> ```
> 
> That would map pretty easily into what we're doing in the [hostlist](https://github.com/daos-stack/daos/tree/master/src/control/lib/hostlist) package, and should be easier to work with on the C side than strings.

Indeed, the idea was to not transfer formatted string through protobuf.
I was also thinking to transfer such protobuf struct (close to what is used in the engine) which are decoupled to the end user string format representation.

Reply:
Keeping this simple and transferring `uint32 enabled_ranks` over protobuf would be my preferred solution if range string is not acceptable. Given the infrequency of the operation would a ~2k message introduce a problem?

Reply:
@knard-intel IMO better to solve this problem in a new PR associated with a new ticket. I think it's OK to leave the strings for now.

Reply:
@tanabarr Fair point... I think it would be a good idea to test and see what the impact is.

Reply:
From my side both solution (integer interval and explicit list of ranks) make sense.
Interval of integer should not be so hard to implement on the engine side as it is already building a list of intervals.
If it is OK for you, I will submit this PR as it for landing (as soon as the CI is OK), and I will continue investigation on this point in a follow-up PR.
If it is not OK for you, do not hesitate to ask for modifications.

Reply:
Follow-up action regarding decoupling improvement will be done in the ticket [DAOS-15987](https://daosio.atlassian.net/browse/DAOS-15987)